### ISSUE TITLE
[Question] Where does non-default keywords fall within the set of `use-package-keywords` in terms of the order they are processed?

### ISSUE THREAD

**Issue by doolio** [jwiegley/use-package#565]
You have defined the `defcustom use-package-keywords` as the set of valid keywords, in the order they are processed. Yet this (default) set does not include `:ensure`, `:pin`, `:diminish` or `:delight` with the recent changes/improvements. I see (for me) from `C-h v` that `:pin` and `:ensure` are inserted to the front of this list and `:diminish` and `:delight` to the back of this list. I take this to mean for example that `:pin` then `:ensure` if added are processed before all other keywords etc. 

Would it be prudent to include these in the `defcustom` as well (say as comments) or to comment on them in its docstring so this is clear to those who examine the code. I raise this as I make use of both `:ensure` and `:pin` as well as `:diminish` in several of my `use-package` forms and wanted to ensure I was respecting the order and so sought out this `defcustom`.

Similarly, was your intention to maintain the order they are processed in the `defcustom use-package-deferring-keywords` set as well then I think `:hook` should appear before `:commands`. Thanks for your time.

**Comment by jwiegley** [comment]
I'm not entirely sure what your comment means, @doolio, can you rephrase?

Opt-in add-on packages add their keywords on load up.  bind-key currently is still special.  I'm currently not documenting add-ons in any of the core docstrings, since the intention is that there will be several more add-ons in the future.

**Comment by doolio** [comment]
I think you've answered my question in any case. I was suggesting including these keywords (from as you've explained these opt-in packages) in the `defcustom` as arguments commented out or by wording in the docstring. Basically, I meant something like the following. Currently, you have:

````
(defcustom use-package-keywords
  '(:disabled
    :load-path
    :requires
    :defines
    :functions
    :preface
    :if :when :unless
    :no-require
    :catch
    :after
    :custom
    :custom-face
    :bind
    :bind*
    :bind-keymap
    :bind-keymap*
    :interpreter
    :mode
    :magic
    :magic-fallback
    :hook
    ;; Any other keyword that also declares commands to be autoloaded (such as
    ;; :bind) must appear before this keyword.
    :commands
    :init
    :defer
    :demand
    :load
    ;; This must occur almost last; the only forms which should appear after
    ;; are those that must happen directly after the config forms.
    :config)
  "The set of valid keywords, in the order they are processed in.
The order of this list is *very important*, so it is only
advisable to insert new keywords, never to delete or reorder
them. Further, attention should be paid to the NEWS.md if the
default order ever changes, as they may have subtle effects on
the semantics of use-package declarations and may necessitate
changing where you had inserted a new keyword earlier.
Note that `:disabled' is special in this list, as it causes
nothing at all to happen, even if the rest of the use-package
declaration is incorrect."
  :type '(repeat symbol)
:group 'use-package)
````

and I was asking whether it made sense to have something like the following to illustrate where these keywords fall wrt the process order so people who queried this definition directly would see it. 

````
(defcustom use-package-keywords
  '(;; :pin       ; from the ensure opt-in package
    ;; :ensure    ; from the ensure opt-in package
    :disabled
    :load-path
    :requires
    :defines
    :functions
    :preface
    :if :when :unless
    :no-require
    :catch
    :after
    :custom
    :custom-face
    :bind
    :bind*
    :bind-keymap
    :bind-keymap*
    :interpreter
    :mode
    :magic
    :magic-fallback
    :hook
    ;; Any other keyword that also declares commands to be autoloaded (such as
    ;; :bind) must appear before this keyword.
    :commands
    :init
    :defer
    :demand
    :load
    ;; This must occur almost last; the only forms which should appear after
    ;; are those that must happen directly after the config forms.
    :config
    ;; :diminish  ; from the diminish opt-in package
    ;; :delight   ; from the delight opt-in package
   ) 
  "The set of valid keywords, in the order they are processed in.
The order of this list is *very important*, so it is only
advisable to insert new keywords, never to delete or reorder
them. Further, attention should be paid to the NEWS.md if the
default order ever changes, as they may have subtle effects on
the semantics of use-package declarations and may necessitate
changing where you had inserted a new keyword earlier.
Note that `:disabled' is special in this list, as it causes
nothing at all to happen, even if the rest of the use-package
declaration is incorrect."
  :type '(repeat symbol)
:group 'use-package)
````

but I see now that this doesn't really add value and is also not very attractive particularly if you plan on adding many more add-ons in the future. The existing docstring already emphasises that the list is in the order the keywords are processed. Therefore, I think we can easily close this issue.
